design/gas/Design_gas.md

```diff
@@ -0,0 +1,89 @@
+2. Bridged gas should automatically deposit into the respective addresses.
+3. Gas-consuming transactions should include a layer 1 cost component for calldata publishing and a layer 2 component that covers execution costs and prevents endless transaction execution.
+4. L2 gas price should be configurable. Furthermore it should support being free whilst still not allowing non halting execution.
+5. The gas mechanics need to prevent denial of service attacks.
+6. EIP-1559 transactions should have their base fee + tip deposited to the sequencer without any ETH being burned.
+
+## Gas Cost Display
+
+Metamask can disassemble the cost components for layer 1 and layer 2 for specific transactions on Optimism, such as token balance approvals. However, it remains unclear how they achieve this as it's neither fully documented nor supports all transactions.
+
+
+## Design for Gas Deposit
+
+Given that Ethereum as a currency operates on the layer 1 protocol, it needs to be bridged to Obscuro for usage. This transition mechanism will be facilitated by the standard Obscuro bridge. The deposited value will produce a cross-chain message indicating the Ethereum recipient through the bridge smart contract.
+
+```EthereumDeposited(address receiver, uint256 amount)```
+
+The enclave will automatically pick up this message, along with other cross-chain messages, but it will be treated specially. Upon detection, the enclave will increase the balance of the receiver account by the amount specified in the message. This automatic relaying is needed in order to avoid the chicken and egg problem one would have initially where no relayer has any gas to relay the gas deposit messages. It would also ensure relayers who run out of gas can easily recharge without relying on other relayers if we were to use a different bootstrap mechanism. Note that those deposits will not call the fallback function of the address if it is a smart contract and any deposits to such contracts would later on need a message relayed to verify the message as they would normally.
+
+## Gas Estimation
+
+When an RPC call for gas estimation is made to Obscuro, the returned estimate should include the expected layer 1 calldata cost alongside the usual execution-based gas estimation. This can be achieved by taking an estimation for layer 1 minFee per gas, determining the calldata cost of the transaction, and adding this to the standard gas estimate.
+
+`Gas estimate = L1_gas_fee * calldata_gas + estimated_l2_gas`
+
+## Gas Expenditure
+
+Before executing a transaction at Obscuro, we must automatically deduct the layer 1 costs for the transaction from the sender's balance. This can be done directly through the stateDB before processing. Ideally, the remaining balance will be refunded, simulating a scenario where the gas limit has been spent. However, potential overruns of execution costs could result in transactions exceeding their set gas limit.
```

Comment:
Let's say we have a mechanism to estimate the L1 eth gas cost.
The following happens:

1. Obscuro user submits TX1 (size=1kb) and the L2 execution cost is 1Million gas.
2. Tx1 enters the enclave which will do the following:
   a. Check that the user has enough balance to pay for size(Tx1) * estimated_publishing cost. If yes, deduct that amount. Also, record this amount, in case it needs to be payed back if the tx is not included.
   b. Check the user has  the minium gas required for Obscuro. Deduct that
   c. Start executing, until the tx finishes, or the user runs out of gas.

Reply:
This is the initial approach I was thinking about as it is most intuitive to implement, but the problem with it is that transactions will result in spending more ETH than authorised when signed. When a user signs there is an agreement over the total gas cost that can be deducted from the balance + the total value being sent in the transaction; Deducting externally of the gas will result in higher than authorized "total" gas spend.

Reply:
but why do you have to deduct more than authorised?

Obscuro user submits TX1 (size=1kb) and the L2 execution cost is 1Million gas.

Tx1 enters the enclave which will do the following:

a. Check that the user has enough balance to pay for size(Tx1) * estimated_publishing cost. Also if the tx is authorised to spend that. If yes, deduct that amount. Also, record this amount, in case it needs to be payed back if the tx is not included.
b. Check the user has the minium gas required for Obscuro. Deduct that
c. Start executing, until the tx finishes, or the user runs out of gas.